test/expr/expressions.swift

```diff
@@ -702,6 +707,12 @@ func unusedExpressionResults() {
   let optionalc:C? = nil
   optionalc?.method()  // ok
   optionalc?.method  // expected-error {{expression resolves to an unused function}}
+
+  optionalc?.method2 // expected-error {{expression resolves to an unused function}}
+  optionalc?.method2() // expected-error {{expression resolves to an unused function}}
+
+  optionalc?.method3 // ok
```

Comment:
This one isn't right; this *is* an unused function, not an unused result.

Reply:
Oh? This currently (am using v4.1) passes tho? Is it a bug then?

Reply:
Yes, that's a bug too! Whoops.

Reply:
Representing methods as `(A) -> () -> String` has so many unfortunate implications…!

Reply:
Let me confirm so everything gets clear before fixing.

```swift
final class A {
  func method1() -> String {
    return ""
  }
}

// Expected behaviours 
A.method1 // expression resolves to an unused function
A().method1 // expression resolves to an unused function


```

Reply:
Let me confirm so everything gets clear before fixing:

```swift
final class A {
  func method1() -> String {
    return ""
  }

  @discardableResult
  func method2() -> String {
    return ""
  }

  @discardableResult
  func method3() -> () -> String {
    return { "" }
  }
}


// Expected
A.method1   // expression resolves to an unused function
A().method1 // expression resolves to an unused function
A.method2   // expression resolves to an unused function
A.method3   // expression resolves to an unused function

// Bugs
A().method2 // passes -> expression resolves to an unused function
A().method3() // expression resolves to an unused function -> passes
```

Reply:
Yes, that looks correct to me. If it's hard to fix the `A.methodN(A())` cases it's probably okay to have those pass even if they really shouldn't; doing that is rare anyway.

Reply:
New question - what do you think of `A().method3()()`? I feel it should complain about unused result but it does nothing so far..

Reply:
I agree that that one should complain too! It seems rare as well, though.